Src/Base/AMReX_DistributionMapping.cpp

```diff
@@ -1410,6 +1410,183 @@ DistributionMapping::makeKnapSack (const Vector<Real>& rcost, Real& eff, int nma
+        Vector<int> old_index_to_new_index;
+        old_index_to_new_index.resize(rcost_local.size(), 0);
+        
+        for (int i=0; i<old_index_to_new_index.size(); i++)
+        {
+            old_index_to_new_index[new_index_to_old_index[i]] = i;
+        }
+        
+        for (int i=0; i<rcost.size(); i++)
+        {
+            rcost[i] = new_index_to_cost[old_index_to_new_index[i]];
+        }
+    }
+#else
+    for (int i=0; i<rcost_local.size(); i++)
+    {
+        rcost[i] = rcost_local[i];
+    }
+#endif
+
+    // rcost is filled out on root
+    if (ParallelDescriptor::MyProc() == root)
+    {
+        Vector<Long> cost(rcost.size());
+
+        Real wmax = *std::max_element(rcost.begin(), rcost.end());
+        Real scale = (wmax == 0) ? 1.e9 : 1.e9/wmax;
+
+        for (int i = 0; i < rcost.size(); ++i) {
+            cost[i] = Long(rcost[i]*scale) + 1L;
+        }
+
+        int nprocs = ParallelContext::NProcsSub();
+
+        // `sort` needs to be false here since there's a parallel reduce function
+        // in the processor map function, but we are executing only on root 
+        r.KnapSackProcessorMap(cost, nprocs, &proposedEfficiency, true, nmax, false);
+
+        ComputeDistributionMappingEfficiency(rcost_local.DistributionMap(),
+                                             rcost,
+                                             &currentEfficiency);
+    }
+
+#ifdef BL_USE_MPI
+    // Load-balanced distribution mapping is computed on root; broadcast the cost
+    // to all proc (optional)
+    if (broadcastToAll)
+    {
+        Vector<int> pmap;
+        pmap.resize(rcost_local.DistributionMap().size(), 0);
+        pmap.assign(rcost_local.DistributionMap().size(), 0);
+        if (ParallelDescriptor::MyProc() == root)
+        {
+            pmap = r.ProcessorMap();
+        }
+        
+        // Broadcast vector from which to construct new distribution mapping
+        ParallelDescriptor::Bcast(&pmap[0], pmap.size(), root);
+        if (ParallelDescriptor::MyProc() != root)
+        {
+            r = DistributionMapping(pmap);
+        }
+    }
+#endif
+    
+    return r;
+}
+
+void
+DistributionMapping::ComputeDistributionMappingEfficiency (const DistributionMapping& dm,
+                                                           const Vector<Real>& cost,
+                                                           Real* efficiency)
+{
+    const Real nprocs = ParallelDescriptor::NProcs();
+
+    // This will store mapping from processor to the costs of FABs it controls,
+    // (proc) --> ([cost_FAB_1, cost_FAB_2, ... ]),
+    // for each proc
+    std::map<int, Vector<Real>> rankToCosts;
+
+    for (int i=0; i<cost.size(); ++i)
+    {
+        rankToCosts[dm[i]].push_back(cost[i]);
+    }
+
+    Real maxCost = -1.0;
+
+    // This will store mapping from (proc) --> (sum of cost) for each proc
+    Vector<Real> rankToCost = {0.0};
+    rankToCost.resize(nprocs);
```

Comment:
It seems that this can be `Vector<Real> rankToCost(nprocs);`.

Reply:
Changed to this.